File: frontend/src/api/finance.js
/** 理财助手 API */

import { api } from "./http.js";

/** ── 市场指数 ── */

export function fetchMarketIndices() {
  return api("/api/v1/finance/market-indices");
}

/** ── A 股 ── */

export function searchStocks(query) {
  return api(`/api/v1/finance/stock/search?q=${encodeURIComponent(query)}`);
}

export function fetchStockQuotes(codes) {
  return api(
    `/api/v1/finance/stock/quote?codes=${encodeURIComponent(codes.join(","))}`
  );
}

export function fetchStockKline(code, ktype = "day") {
  return api(`/api/v1/finance/stock/kline?code=${code}&ktype=${ktype}`);
}

/** ── 基金 ── */

export function searchFunds(query) {
  return api(`/api/v1/finance/fund/search?q=${encodeURIComponent(query)}`);
}

export function fetchFundQuote(code) {
  return api(`/api/v1/finance/fund/quote?code=${code}`);
}

export function fetchFundHistory(code, period = "3m") {
  return api(
    `/api/v1/finance/fund/history?code=${code}&period=${period}`
  );
}

/** ── 虚拟币 ── */

export function fetchCryptoList(perPage = 50) {
  return api(`/api/v1/finance/crypto/list?per_page=${perPage}`);
}

export function fetchCryptoQuote(coinId) {
  return api(`/api/v1/finance/crypto/quote?coin_id=${encodeURIComponent(coinId)}`);
}

export function fetchCryptoHistory(coinId, days = 7) {
  return api(
    `/api/v1/finance/crypto/history?coin_id=${encodeURIComponent(coinId)}&days=${days}`
  );
}

/** ── 自选清单 ── */

export function fetchWatchlist() {
  return api("/api/v1/finance/watchlist");
}

export function addToWatchlist(assetType, assetCode, assetName) {
  return api("/api/v1/finance/watchlist", {
    method: "POST",
    body: JSON.stringify({ asset_type: assetType, asset_code: assetCode, asset_name: assetName }),
  });
}

export function removeFromWatchlist(itemId) {
  return api(`/api/v1/finance/watchlist/${itemId}`, { method: "DELETE" });
}

/** ── 报告任务 ── */

export function submitReport(payload) {
  return api("/api/v1/finance/report", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function fetchReports(status = "", limit = 50, offset = 0) {
  const params = new URLSearchParams();
  if (status) params.set("status", status);
  params.set("limit", limit);
  params.set("offset", offset);
  return api(`/api/v1/finance/reports?${params.toString()}`);
}

export function cancelReport(reportId) {
  return api(`/api/v1/finance/report/${reportId}/cancel`, { method: "POST" });
}

export function fetchReportDetail(reportId) {
  return api(`/api/v1/finance/report/${reportId}`);
}

/** 公开分享链接（无需登录即可打开） */
export function getReportShareUrl(shareToken) {
  if (!shareToken) return "";
  // 与 http.js 默认一致：浏览器侧走同源 /ai 反代
  const base = "/ai";
  const origin = typeof window !== "undefined" ? window.location.origin : "";
  return `${origin}${base}/api/v1/share/finance/${shareToken}`;
}

/** 在线查看报告：打开可独立访问的公开分享 URL */
export async function viewReport(reportId, shareToken) {
  const { openExternal } = await import("../utils/openExternal.js");
  const { getApiBase, getToken } = await import("./http.js");

  let token = shareToken;
  if (!token) {
    try {
      const detail = await fetchReportDetail(reportId);
      token = detail?.share_token || detail?.data?.share_token;
    } catch {
      /* ignore */
    }
  }
  if (token) {
    openExternal(getReportShareUrl(token));
    return;
  }

  // 兜底：登录态跳转接口（会 302 到公开链接）
  const auth = getToken();
  const base = (getApiBase() || "/ai").replace(/\/$/, "");
  const url = `${window.location.origin}${base}/api/v1/finance/report/${reportId}/view`;
  if (auth) {
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${auth}` },
      redirect: "manual",
    });
    const loc = res.headers.get("Location");
    if (loc) {
      openExternal(loc.startsWith("http") ? loc : `${window.location.origin}${loc}`);
      return;
    }
  }
  openExternal(url);
}

export async function deleteReport(reportId) {
  return api(`/api/v1/finance/report/${reportId}`, { method: "DELETE" });
}

/** 将已完成报告加入个人级文档库（未分类） */
export function importReportToLibrary(reportId) {
  return api(`/api/v1/finance/report/${reportId}/import-library`, {
    method: "POST",
  });
}

/** 生成/覆盖报告公开分享令牌 */
export function shareReport(reportId, { regenerate = true } = {}) {
  const q = regenerate ? "?regenerate=true" : "?regenerate=false";
  return api(`/api/v1/finance/report/${reportId}/share${q}`, { method: "POST" });
}

/** 撤销报告公开分享链接 */
export function unshareReport(reportId) {
  return api(`/api/v1/finance/report/${reportId}/share`, { method: "DELETE" });
}

export async function downloadReport(reportId, fmt = "md") {
  const { getApiBase, getToken } = await import("./http.js");
  const { default: downloadBlob } = await import("../utils/downloadBlob.js");
  const base = getApiBase();
  const token = getToken();
  const res = await fetch(
    `${base}/api/v1/finance/report/${reportId}/download?fmt=${fmt}`,
    { headers: token ? { Authorization: `Bearer ${token}` } : {} }
  );
  if (!res.ok) throw new Error("下载失败");
  const blob = await res.blob();
  const ext = fmt === "docx" ? ".docx" : fmt === "pdf" ? ".pdf" : ".md";
  downloadBlob(blob, `report_${reportId.slice(0, 8)}${ext}`);
}

/** 导出报告为 PDF（打开公开页后触发打印） */
export async function exportReportPdf(reportId, shareToken) {
  const { openExternal } = await import("../utils/openExternal.js");
  let token = shareToken;
  if (!token) {
    try {
      const detail = await fetchReportDetail(reportId);
      token = detail?.share_token || detail?.data?.share_token;
    } catch {
      /* ignore */
    }
  }
  if (!token) throw new Error("缺少分享链接");
  // 打开公开页，用户可在浏览器中打印为 PDF
  openExternal(getReportShareUrl(token));
}

File: backend/app/api/finance.py
"""理财助手 API — A股/基金/虚拟币行情接口。"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import HTMLResponse, PlainTextResponse, RedirectResponse
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, require_feature
from app.database import get_db
from app.models.finance_report import FinanceReport
from app.models.org import User
from app.schemas.common import ApiResponse
from app.schemas.finance import (
    ReportImportLibraryOut,
    ReportOut,
    ReportSubmit,
    WatchlistItemCreate,
    WatchlistItemOut,
)
from app.services import finance_service as svc

router = APIRouter(
    prefix="/finance",
    tags=["finance"],
    dependencies=[Depends(require_feature("finance_assistant"))],
)

# 公开分享路由：无登录、无功能权限依赖
public_router = APIRouter(prefix="/share/finance", tags=["finance-share"])


def _report_title(r: FinanceReport) -> str:
    title = f"「{r.stock_name} ({r.stock_code})」"
    if r.report_type == "roundtable":
        direction = "基本面" if r.research_direction == "fundamental" else "短线"
        kind = "辩论版" if r.roundtable_type == "debate" else "专业研究"
        title = f"「{r.stock_name} ({r.stock_code})」{direction}圆桌 · {kind}"
    elif r.report_type == "ai":
        title = f"「{r.stock_name} ({r.stock_code})」AI 深度解读"
    elif r.report_type == "vpa":
        title = f"「{r.stock_name} ({r.stock_code})」量价会诊"
    return title


def _render_completed_report(db: Session, r: FinanceReport) -> HTMLResponse:
    from app.services.finance_report_render import render_report_html

    try:
        r.view_count = int(r.view_count or 0) + 1
        db.commit()
        db.refresh(r)
    except Exception:
        db.rollback()

    html = render_report_html(
        r.content or "",
        title=_report_title(r),
        created_at=r.created_at,
        completed_at=r.completed_at,
        view_count=int(r.view_count or 0),
    )
    return HTMLResponse(content=html)


# ── A 股 ───────────────────────────────────────────────────────


@router.get("/market-indices", response_model=ApiResponse)
async def market_indices(
    _: Annotated[User, Depends(get_current_user)],
) -> ApiResponse:
    """主要市场指数行情（上证、深证、创业板等）。"""
    data = await svc.get_market_indices()
    return ApiResponse(data=data)


@router.get("/stock/search", response_model=ApiResponse)
async def stock_search(
    _: Annotated[User, Depends(get_current_user)],
    q: str = Query("", description="搜索关键词"),
) -> ApiResponse:
    """搜索 A 股（代码或名称）。"""
    data = await svc.search_stocks(q)
    return ApiResponse(data=data)


@router.get("/stock/quote", response_model=ApiResponse)
async def stock_quote(
    _: Annotated[User, Depends(get_current_user)],
    codes: str = Query("", description="股票代码，逗号分隔"),
) -> ApiResponse:
    """批量获取 A 股实时行情。"""
    code_list = [c.strip() for c in codes.split(",") if c.strip()]
    data = await svc.get_stock_quotes(code_list)
    return ApiResponse(data=data)


@router.get("/stock/kline", response_model=ApiResponse)
async def stock_kline(
    _: Annotated[User, Depends(get_current_user)],
    code: str = Query(..., description="股票代码"),
    ktype: str = Query("day", description="K 线类型: day/week/month"),
) -> ApiResponse:
    """获取 A 股 K 线数据。"""
    data = await svc.get_stock_kline(code, ktype)
    return ApiResponse(data=data)


# ── 基金 ───────────────────────────────────────────────────────


@router.get("/fund/search", response_model=ApiResponse)
async def fund_search(
    _: Annotated[User, Depends(get_current_user)],
    q: str = Query("", description="搜索关键词"),
) -> ApiResponse:
    """搜索基金（代码或名称）。"""
    data = await svc.search_funds(q)
    return ApiResponse(data=data)


@router.get("/fund/quote", response_model=ApiResponse)
async def fund_quote(
    _: Annotated[User, Depends(get_current_user)],
    code: str = Query(..., description="基金代码"),
) -> ApiResponse:
    """获取基金实时估值 / 最新净值。"""
    data = await svc.get_fund_quote(code)
    return ApiResponse(data=data)


@router.get("/fund/history", response_model=ApiResponse)
async def fund_history(
    _: Annotated[User, Depends(get_current_user)],
    code: str = Query(..., description="基金代码"),
    period: str = Query("3m", description="周期: 1m/3m/6m/1y/all"),
) -> ApiResponse:
    """获取基金历史净值。"""
    data = await svc.get_fund_history(code, period)
    return ApiResponse(data=data)


# ── 虚拟币 ─────────────────────────────────────────────────────


@router.get("/crypto/list", response_model=ApiResponse)
async def crypto_list(
    _: Annotated[User, Depends(get_current_user)],
    per_page: int = Query(50, description="数量"),
) -> ApiResponse:
    """Top 虚拟币行情。"""
    data = await svc.list_crypto(per_page)
    return ApiResponse(data=data)


@router.get("/crypto/quote", response_model=ApiResponse)
async def crypto_quote(
    _: Annotated[User, Depends(get_current_user)],
    coin_id: str = Query(..., description="CoinGecko coin id"),
) -> ApiResponse:
    """获取单个虚拟币详情。"""
    data = await svc.get_crypto_quote(coin_id)
    return ApiResponse(data=data)


@router.get("/crypto/history", response_model=ApiResponse)
async def crypto_history(
    _: Annotated[User, Depends(get_current_user)],
    coin_id: str = Query(..., description="CoinGecko coin id"),
    days: int = Query(7, description="天数: 1/7/30/365"),
) -> ApiResponse:
    """获取虚拟币历史价格。"""
    data = await svc.get_crypto_history(coin_id, days)
    return ApiResponse(data=data)


# ── 自选清单 ─────────────────────────────────────────────────────


@router.get("/watchlist", response_model=ApiResponse)
def list_watchlist(
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse:
    """获取用户的理财自选清单。"""
    items = svc.list_watchlist(db, user.id)
    return ApiResponse(data=[WatchlistItemOut.model_validate(i) for i in items])


@router.post("/watchlist", response_model=ApiResponse[WatchlistItemOut])
def add_watchlist(
    body: WatchlistItemCreate,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[WatchlistItemOut]:
    """添加自选。"""
    item = svc.add_watchlist(
        db,
        user.id,
        asset_type=body.asset_type,
        asset_code=body.asset_code,
        asset_name=body.asset_name,
    )
    return ApiResponse(data=WatchlistItemOut.model_validate(item))


@router.delete("/watchlist/{item_id}", response_model=ApiResponse)
def remove_watchlist(
    item_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse:
    """删除自选。"""
    ok = svc.remove_watchlist(db, user.id, item_id)
    return ApiResponse(data={"deleted": ok})


# ── 报告任务 ─────────────────────────────────────────────────────


@router.post("/report", response_model=ApiResponse[ReportOut])
async def submit_report(
    body: ReportSubmit,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[ReportOut]:
    """提交理财报告生成任务（异步执行）。"""
    report = svc.create_report(
        db,
        user.id,
        stock_code=body.stock_code,
        stock_name=body.stock_name,
        report_type=body.report_type,
        roundtable_type=body.roundtable_type,
        research_direction=body.research_direction,
        ai_context=body.ai_context,
    )
    await svc.submit_report_task(report)
    return ApiResponse(data=ReportOut.model_validate(report))


@router.get("/reports", response_model=ApiResponse[list[ReportOut]])
def list_reports(
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
    status: str = Query("", description="过滤状态: pending/running/completed/failed"),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
) -> ApiResponse[list[ReportOut]]:
    """获取用户报告列表。"""
    q = status.strip() or None
    items = svc.get_user_reports(db, user.id, status=q, limit=limit, offset=offset)
    return ApiResponse(data=[ReportOut.model_validate(i) for i in items])


@router.get("/report/{report_id}", response_model=ApiResponse[ReportOut])
def get_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[ReportOut]:
    """获取单个报告详情。"""
    r = svc.get_report(db, report_id)
    if not r or r.user_id != user.id:
        raise HTTPException(status_code=404, detail="报告不存在")
    return ApiResponse(data=ReportOut.model_validate(r))


@router.post("/report/{report_id}/cancel", response_model=ApiResponse[ReportOut])
def cancel_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[ReportOut]:
    """取消报告生成任务。"""
    try:
        report = svc.cancel_report_task(db, user.id, report_id)
        return ApiResponse(data=ReportOut.model_validate(report))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/report/{report_id}/view")
def view_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
):
    """登录态查看：跳转到公开分享链接（可单独打开、可转发）。"""
    r = svc.get_report(db, report_id)
    if not r or r.user_id != user.id:
        raise HTTPException(status_code=404, detail="报告不存在")
    if r.status != "completed" or not r.content:
        raise HTTPException(status_code=400, detail="报告尚未完成")
    token = svc.ensure_share_token(db, r)
    return RedirectResponse(url=f"/api/v1/share/finance/{token}", status_code=302)


@router.post("/report/{report_id}/share", response_model=ApiResponse[ReportOut])
def share_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
    regenerate: bool = Query(True, description="是否重新生成令牌（覆盖旧链接）"),
) -> ApiResponse[ReportOut]:
    """生成/覆盖报告公开分享令牌。"""
    r = svc.get_report(db, report_id)
    if not r or r.user_id != user.id:
        raise HTTPException(status_code=404, detail="报告不存在")
    if r.status != "completed" or not r.content:
        raise HTTPException(status_code=400, detail="报告尚未完成")
    if regenerate or not r.share_token:
        svc.regenerate_share_token(db, r)
    else:
        svc.ensure_share_token(db, r)
    return ApiResponse(data=ReportOut.model_validate(r))


@router.delete("/report/{report_id}/share", response_model=ApiResponse[ReportOut])
def unshare_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[ReportOut]:
    """撤销报告公开分享链接。"""
    r = svc.get_report(db, report_id)
    if not r or r.user_id != user.id:
        raise HTTPException(status_code=404, detail="报告不存在")
    svc.revoke_share_token(db, r)
    return ApiResponse(data=ReportOut.model_validate(r))


@public_router.get("/{share_token}", response_class=HTMLResponse)
def view_shared_report(
    share_token: str,
    db: Annotated[Session, Depends(get_db)],
) -> HTMLResponse:
    """公开分享页：无需登录即可查看已完成报告。"""
    r = svc.get_report_by_share_token(db, share_token)
    if not r or r.status != "completed" or not r.content:
        raise HTTPException(status_code=404, detail="报告不存在或尚未完成")
    return _render_completed_report(db, r)


@router.get("/report/{report_id}/download")
def download_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
    fmt: str = Query("md", description="格式: md/pdf/docx"),
) -> PlainTextResponse:
    """下载报告内容。

    fmt 支持:
      - md:   Markdown 原文（默认）
      - pdf:  暂不支持（返回 md 占位）
      - docx: 暂不支持（返回 md 占位）
    """
    r = svc.get_report(db, report_id)
    if not r or r.user_id != user.id:
        raise HTTPException(status_code=404, detail="报告不存在")
    if r.status != "completed" or not r.content:
        raise HTTPException(status_code=400, detail="报告尚未完成")

    content = r.content
    filename = f"{r.stock_code}_{r.report_type}_{r.created_at.strftime('%Y%m%d')}"

    if fmt == "pdf":
        # TODO: 接入 PDF 渲染引擎
        return PlainTextResponse(
            content=content,
            media_type="text/markdown",
            headers={
                "Content-Disposition": f'attachment; filename="{filename}.md"',
            },
        )
    if fmt == "docx":
        # TODO: 接入 DOCX 渲染引擎
        return PlainTextResponse(
            content=content,
            media_type="text/markdown",
            headers={
                "Content-Disposition": f'attachment; filename="{filename}.md"',
            },
        )

    # 默认 Markdown
    return PlainTextResponse(
        content=content,
        media_type="text/markdown",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}.md"',
        },
    )


@router.post(
    "/report/{report_id}/import-library",
    response_model=ApiResponse[ReportImportLibraryOut],
)
def import_report_to_library(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[ReportImportLibraryOut]:
    """将已完成报告加入个人级文档库（默认未分类）。"""
    data = svc.import_report_to_library(db, user, report_id)
    return ApiResponse(data=ReportImportLibraryOut.model_validate(data))


@router.delete("/report/{report_id}", response_model=ApiResponse)
def delete_report(
    report_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse:
    """删除报告。"""
    ok = svc.delete_report(db, user.id, report_id)
    if not ok:
        raise HTTPException(status_code=404, detail="报告不存在")
    return ApiResponse(data={"deleted": True})
